# IC Data Consistency Checks

## This notebook contains the following topics
### 01. Importing Libraries and Data
### 02. Data Consistency Checks for products.csv
##### a) Mixed Type Data
##### b) Missing Values
##### c) Duplicates
### 03. Data Consistency Checks for wrangled_orders.csv

# 01. Importing Libraries and Data

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Create folder path
path = r'C:\Users\ezrela\Documents\CareerFoundry\DD-MM-2020 Instacart Basket Analysis'

In [4]:
path

'C:\\Users\\ezrela\\Documents\\CareerFoundry\\DD-MM-2020 Instacart Basket Analysis'

In [6]:
# Import products.csv as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [7]:
# Import orders_wrangled.csv as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 02. Data Consistency Checks for products.csv

## Mixed Type Data

In [9]:
# Create a dataframe
df_test = pd.DataFrame()

In [10]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [16]:
# Correcting mixed data type by changing all data to string
df_test['mix'] = df_test['mix'].astype

In [17]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

## Missing Values

In [18]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [22]:
# Creating a subset of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [23]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [24]:
# Shape of original dataframe
df_prods.shape

(49693, 5)

In [25]:
# Dropping the column with missing values by creating a subset without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [26]:
# Shape of dataframe after dropping the missing values
df_prods_clean.shape

(49677, 5)

##  Duplicates

In [28]:
# Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [33]:
# Addressing duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [34]:
df_prods_clean_no_dups.shape

(49672, 5)

In [35]:
# Exporting df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 03. Data Consistency Checks for wrangled_orders.csv

### Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [36]:
# Getting the descriptive analysis of wrangled_orders.csv file
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

##### Nothing about the data looks off or unusual. The columns of particular interest are the order_day_of_week, order_hour_of_day, and days_since_prior_order. The recorded numbers fall within the range of expected numbers (7 days in a week, 24 hours in a day)

### Check for mixed-type data in your df_ords dataframe.

In [38]:
# Checking for mixed-type data
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1) 
    if len (df_ords[weird])>0:
        print (col)

##### There are no mixed data type in df_ords

### Run a check for missing values in your df_ords dataframe.

In [39]:
#Finding missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### The only column with missing values is the column 'days_since_prior_order'. The missing values could be as a result of orders from new customers who do not have prior orders.

In [72]:
# Creating a subset of the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [74]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  new_user_order  
0                        8                     NaN            True  
11                      11                     NaN            True  
26                      14                     NaN            True  
39                      11                     NaN            True  
45                      12                     NaN            True  
...                    ...                     ...             ...  
3420930                 12                     NaN            True  
3420934                 18                     NaN            True  
3421002                 19                     NaN            True  
3421019                 15                     NaN            True  
3421069                 11                     NaN            True  

[206209 rows x 8 columns]

In [43]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

##### The tables above confirm that the missing values are a result of orders from new customers. The 'order_number' column has 1 for all the rows with 'NaN' in the 'days_since_prior_order' column, indicating that this is the first order from respective user. The rows without missing values have order_numbers above 1.

### Address the missing values using an appropriate method.

In [67]:
# Addressing missing values in df_ords by creating a new flag variable
df_ords['new_user_order'] = df_ords['days_since_prior_order'].isnull()

In [68]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_prior_order  new_user_order  
0                  8                     NaN            True  
1                  7                    15.0           False  
2                 12                    21.0           False  
3                  7                    29.0           False  
4                 15                    28.0           False

In [71]:
df_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

##### I think the most appropraite approach to treat the missing values is to create a new variable, 'new_user_order', indicating that the rows with missing values in the column 'days_since_prior_order' are new users. Inputting values or removing these records will significantly change the true analytical results of the data. Since a new column is being created, and can be dropped, the df_ords was overwritten, instead of creating an entirely new df.

## Run a check for duplicate values in your df_ords data.

In [69]:
# Checking for duplicates in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [70]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, new_user_order]
Index: []

##### There are no duplicated rows in df_ords

In [75]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [76]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

##### The maximum price in the products.csv looks like a typographical error in the price column of the file. The value of the mean also shows that the maximum price is unusually high; therefore, it is an anomaly. Also, the maximum product_id may be an anomaly as the expected max product_id is 49693, based on the count. This could indicate duplicated product_id, that is, a product_id appearing more than once.